osu.Server.Spectator/Database/DatabaseAccess.cs

```diff
@@ -154,22 +156,40 @@ public async Task UpdateRoomParticipantsAsync(MultiplayerRoom room)
             }
         }
 
-        public async Task ClearRoomScoresAsync(MultiplayerRoom room)
+        public async Task<multiplayer_playlist_item> GetPlaylistItemFromRoomAsync(long roomId, long playlistItemId)
         {
-            // for now, clear all existing scores out of the playlist item to ensure no duplicates.
-            // eventually we will want to increment to a new playlist item rather than reusing the same one.
-            long playlistItemId = await connection.QuerySingleAsync<long>("SELECT id FROM multiplayer_playlist_items WHERE room_id = @RoomID", new
+            return await connection.QueryFirstOrDefaultAsync<multiplayer_playlist_item>("SELECT * FROM multiplayer_playlist_items WHERE id = @Id AND room_id = @RoomId", new
             {
-                RoomID = room.RoomID,
+                Id = playlistItemId,
+                RoomId = roomId
             });
+        }
+
+        public async Task<long> AddPlaylistItemAsync(multiplayer_playlist_item item)
+        {
+            await connection.ExecuteAsync(
+                "INSERT INTO multiplayer_playlist_items (room_id, beatmap_id, ruleset_id, allowed_mods, required_mods, created_at, updated_at)"
+                + " VALUES (@room_id, @beatmap_id, @ruleset_id, @allowed_mods, @required_mods, NOW(), NOW())",
+                item);
 
-            await connection.ExecuteAsync("DELETE FROM multiplayer_scores WHERE playlist_item_id = @PlaylistItemID", new { PlaylistItemID = playlistItemId });
-            await connection.ExecuteAsync("DELETE FROM multiplayer_scores_high WHERE playlist_item_id = @PlaylistItemID", new { PlaylistItemID = playlistItemId });
+            return await connection.QuerySingleAsync<long>("SELECT max(id) FROM multiplayer_playlist_items WHERE room_id = @room_id", item);
         }
 
-        public Task EndMatchAsync(MultiplayerRoom room)
+        public async Task ExpirePlaylistItemAsync(long playlistItemId)
         {
-            return connection.ExecuteAsync("UPDATE multiplayer_rooms SET ends_at = NOW() WHERE id = @RoomID", new
+            await connection.ExecuteAsync("UPDATE multiplayer_playlist_items SET expired = 1, updated_at = NOW() WHERE id = @PlaylistItemId", new
+            {
+                PlaylistItemId = playlistItemId
+            });
+        }
+
+        public async Task EndMatchAsync(MultiplayerRoom room)
+        {
+            // Remove all non-expired items from the playlist as they have no scores.
```

Comment:
I'm not sure that is correct? Playlist items are expired by `selectNextPlaylistItem()`, which in turn fires on the room transitioning to `Playing` state. I imagine this could therefore discard the scores associated with the last playlist item before the room closes upon all users leaving it.

Reply:
`expired = 1` is set from `Playing` -> `FinishedPlay`, but I think you're right in two cases:
1. Players start play, then all players leave. Technically score submission hasn't occurred yet but the `multiplayer_scores` table has been populated.
2. Players start play, some submit scores, then _all_ players leave. I'm not sure in this case if the room will ever transition to the `FinishedPlay` state before it's closed...

Need more opinions here... Should this even be a thing? The only reason I have this is for visualising the progression in the room - the lounge would otherwise show 5 maps where only 4 have been played, and similarly for something like match history (mp links).

Reply:
What I could do is set the expired state on match start, and keep the select-next part happening on match end. The question I have then is how to handle the client-side visual part of this, where the API has no concept of the "current" playlist item and hides all items with `expired = 1`.

Reply:
The two cases you mention are probably best handled in a separate PR/change. The multiplayer server should be responsible for deciding if a match was "completed" or not, based on whether any single score was submitted, rather than requiring all scores to submit (which I presume from the above description is what is happening right now?).

Basically there would be some method which is run on standard flow (finished play) but also on room close, which decides what to do about any ongoing play, marking it as finished etc.

From a database perspective, marking as `expired` after the play ends sounds correct.

Reply:
> rather than requiring all scores to submit

Yeah, that's correct. Right now it'll nuke the playlist item if the room never transitions to `FinishedPlay`, which can happen via the two cases I mentioned. It actually won't nuke scores since deletes don't cascade, but it still breaks the DB relations (score can point to a playlist item that has been deleted).

So the question is then, should the server do this deletion at all?

Reply:
Only if no scores were submitted. Or never works fine, too. Since the play did start.

Reply:
In that case I'd change this to `DELETE FROM multiplayer_playlist_items p WHERE p.room_id = @room_id AND p.expired = 0 AND (SELECT COUNT(*) FROM multiplayer_scores s WHERE s.playlist_item_id = p.id) = 0`. Sound good?

Reply:
The logic sounds fine, but I'm not sure about how efficient this execution will be. I think it will do for now, but down the line we will probably want to decide which items to delete locally (the server should have the knowledge I think?) rather than doing a nested query.